src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -152,16 +153,24 @@ private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Ur
                 }
             }
 
+            HttpHeaderValueCollection<ProductHeaderValue> upgradeHeader = request.Headers.Upgrade;
```

Comment:
Can we check HasHeaders before accessing Headers, to avoid lazily initializing it if no headers were set?

Reply:
Why is it necessary to check this here?  This gets called on every request and we should keep the logic in this method minimal.  Maybe I don't understand the overall flow here...

Reply:
It seems like we're trying to detect a WebSocket connection here and then convert it to use CONNECT.

I don't think this is the right approach.

We should modify ClientWebSocket to check for a proxied request, and then issue the CONNECT itself instead of a request with an Upgrade header.

There may be some changes necessary to SocketsHttpHandler to enable this, but hopefully the changes are minimal.

Reply:
> We should modify ClientWebSocket to check for a proxied request, and then issue the CONNECT itself instead of a request with an Upgrade header.

At that point, are we getting any benefit from using SocketsHttpHandler? Seems like it'd be basically as easy and cheaper to just use Socket directly, no?

Reply:
The main reason to not use Socket directly is proxy auth.

Reply:
Yeah, I guess that's still worth it.

Reply:
That's really the only reason we use HttpConnection etc internally for establishing an SSL proxy tunnel.  There are some other small advantages, e.g. HttpConnection does all the response parsing for us.  But the main benefit is just proxy auth.

Reply:
The other reason to use SocketsHttpHandler is that just finding a proxy can be difficult, i.e. parsing the system proxy settings (registry keys, environment variables etc.).  A proxy URI can be a static thing or a function of the request URI.  And on Windows, finding a proxy involves using WinHTTP Auto Proxy Discover Service, etc.  So, we want to keep using SocketsHttpHandler since it handles that.

Reply:
Yeah, makes sense.

Reply:
Actually, I'm confused. If CWS is relying on SocketsHttpHandler to find the proxy, how does CWS issue a CONNECT request to the proxy, rather than to the destination url?

Reply:
I don't completely understand the conversation here. I will try to answer the question and summary.

> It seems like we're trying to detect a WebSocket connection here and then convert it to use CONNECT.

Yes, this is my change. For now, we will only use CONNECT for secure connection tunneled through proxy, and this is the codepath for WSS connection. It's intuitive for me to make non-secure websocket connection to use CONNECT here. (I,e,. similar fashion/codepath as WSS).

> We should modify ClientWebSocket to check for a proxied request, and then issue the CONNECT itself instead of a request with an Upgrade header.

I can investigate this approach. This approach is not perfect as well: currently HttpConnection doesn't differentiate websocket connection, to use CONNECT verb (for WSS). 
1. After the change, we may want to isolate websocket connection case from `HttpConnectionPoolManager` class, which still requires header detection/check as well. (cannot avoid the check)
2. Or, choose to only change for non-secure websocket connection case (keep WSS connection codepath unchanged). It's not intuitive because the codepath will be very different, but I can investigate the possibility.

Reply:
My main concern here is that we not have to have logic in critical paths in SocketsHttpHandler that special-cases the ClientWebSocket scenario.

Reply:
Just to expand on this:

I don't think we should try to detect a websocket request in SocketHttpHandler and try to rewrite it.  Logic that is specific to websockets (specifically, whether to do a CONNECT or an upgrade) should be contained within the ClientWebSocket code.

Re this:

> After the change, we may want to isolate websocket connection case from HttpConnectionPoolManager class, which still requires header detection/check as well. (cannot avoid the check)

We already do some checks for (a) successful Upgrade and (b) CONNECT verb.  I don't think we should need additional checks beyond what we have.

When you say "isolate... from HttpConnectionPoolManager", what do you mean exactly?  Are you saying the connection shouldn't be pooled?  It shouldn't be pooled already, I believe.  Or something else?

Reply:
> My main concern here is that we not have to have logic in critical paths in SocketsHttpHandler that special-cases the ClientWebSocket scenario.

I agree with this, but I'm not sure how we avoid this. If ClientWebSocket issues its own CONNECT, it a) needs to know what proxy to connect to, which it won't unless it does its own proxy detection, and b) even if it does know, using SocketsHttpHandler doesn't buy it anything because it can't use the auth support that will end up doing regular auth rather than proxy auth.

So it seems like either:
1) We put something into SocketsHttpHandler to allow ClientWebSocket to just use it.
2) We fore go using SocketsHttpHandler completely and reimplement everything in ClientWebSocket, potentially reusing code files from SocketsHttpHandler.

Or am I missing another simpler solution?

Reply:
Number 1 seems like the only rational choice.

Reply:
I agree we should do 1.  I'm just saying, let's please not introduce invasive, websocket-specific logic into core codepaths.

If we need to, let's just expose SocketsHttpHandler.CreateWebSocketConnection and short-circuit all this stuff.

Reply:
> let's please not introduce invasive, websocket-specific logic into core codepaths.
> If we need to, let's just expose SocketsHttpHandler.CreateWebSocketConnection and short-circuit all this stuff.

Or just allow "ws" and "wss" urls?

Reply:
Also, to address your specific points:

> needs to know what proxy to connect to, which it won't unless it does its own proxy detection

We can just expose the proxy from SocketsHttpHandler.  It's already implemented as an IWebProxy, it's just not exposed publicly.  Expose it and then ClientWebSocket can get it and invoke it.

> even if it does know, using SocketsHttpHandler doesn't buy it anything because it can't use the auth support that will end up doing regular auth rather than proxy auth

That's true, there's no way to say "please do proxy auth on this CONNECT" currently.  Again, we have this internally (we use it for establishing the SSL proxy tunnel) but it's not exposed publicly.

So let's figure out a simple way to expose this so ClientWebSocket can use it.  

One way would be to simply add a method in SocketsHttpHandler like "EstablishProxyTunnel".  That's perhaps a bit ugly, but it gets the job done.

Reply:
> Or just allow "ws" and "wss" urls?

I'm not entirely sure what that means.  Does it mean that SocketsHttpHandler detects these schemes, issues the right Upgrade or CONNECT under the covers, takes the resulting response stream and returns it as the result of the original ws/wss request?  If so I'm fine with that, I think (though I'd like to see the code :)).

Reply:
Ok, 1 makes sense to me. I will redo the PR.

> what do you mean exactly? Are you saying the connection shouldn't be pooled? It shouldn't be pooled already, I believe. Or something else?

Sorry for the confusion, I mean the redundancy logic for WSS to use CONNECT. Your previous comment address my concern: "let's just expose SocketsHttpHandler.CreateWebSocketConnection and short-circuit all this stuff".

Reply:
> Does it mean that SocketsHttpHandler detects these schemes, issues the right Upgrade or CONNECT under the covers, takes the resulting response stream and returns it as the result of the original ws/wss request?

I was thinking more doing exactly what it does today, except a) allow ws and wss (whereas today it throws), but still treat ws as http and wss as https, and b) force tunneling if there's a proxy; otherwise it still leaves everything else up to the caller.  Basically do the stuff that the caller can't, and without exposing additional APIs.

> We can just expose the proxy from SocketsHttpHandler. It's already implemented as an IWebProxy, it's just not exposed publicly. Expose it and then ClientWebSocket can get it and invoke it.

There are two ways this could be done:
1. Initialize SocketsHttpHandler.Proxy to be such an instance rather than null.  I don't think we can use a singleton here, so it would mean forcing all of the work / allocations associated with this instance even if it's not used.
2. Reuse some existing "get me a proxy" public API for doing this.
3. Expose a new API for getting the proxy.

(2)/(3) are https://github.com/dotnet/corefx/issues/26494 I believe.  I'm fine with this, but we're late to design/ship such a thing for 2.1.  Still, if someone could pull this together in a coherent fashion, I'm happy with it.

Reply:
>2.	Reuse some existing "get me a proxy" public API for doing this.

There is already a public API for IWebProxy.GetProxy(Uri destinationUri) method.

In simple cases, a proxy URI is the same for all destination URI's.  But this needs to be called for every new request since it could return a different proxy URI based on the destination URI.

Reply:
> There is already a public API for IWebProxy.GetProxy(Uri destinationUri) method.

I didn't mean "get me a proxy Uri", I meant "get me an IWebProxy that can give me system proxy information".

Reply:
If we do this:

>Initialize SocketsHttpHandler.Proxy to be such an instance rather than null. I don't think we can use a singleton here, so it would mean forcing all of the work / allocations associated with this instance even if it's not used.

then we already have the IWebProxy that computes the right thing based on the default system web proxy settings.  And @wfurt already did the work to call the platform APIs to implement this.

Reply:
> then we already have the IWebProxy that computes the right thing based on the default system web proxy settings

Right, it's an instance of that to which I was referring.  But that's not pay-for-play, which makes me not like it unless we have to.

Reply:
Basically initialize SocketsHttpHandler.Proxy with this:

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L81

Reply:
> Initialize SocketsHttpHandler.Proxy to be such an instance rather than null. I don't think we can use a singleton here

Why can't we use a singleton here?  I think we are not using a singleton today, but I'm not sure that's intentional or desirable.

Reply:
> Why can't we use a singleton here?

Environment variables can change.  If we create a singleton that stores these environment variables, then if they change and the dev wants to respect the change, they're out of luck.  Tying it to the lifetime of the SocketsHttpHandler is at least explainable and consistent.

And WinHttpOpen is, as I understand it, tied to a user, so using a singleton would be problematic in the face of impersonation.

Reply:
Yes, you still need an instance.  But maybe use Lazy init for initializing SocketsHttpHandler.Proxy.  Then if it got overwritten with some custom IWebProxy thing you would save the cost of it.

Reply:
> But maybe use Lazy init for initializing SocketsHttpHandler.Proxy.

That's fair, we could do that.  We probably don't care nearly as much about someone querying Proxy if they don't intend to use it directly.

Reply:
Stephen and I chatted offline.  I think where we landed is what he suggested above:

> I was thinking more doing exactly what it does today, except a) allow ws and wss (whereas today it throws), but still treat ws as http and wss as https, and b) force tunneling if there's a proxy; otherwise it still leaves everything else up to the caller. Basically do the stuff that the caller can't, and without exposing additional APIs.

I believe this addresses my concerns, since we'll detect ws: or wss: once up front and behave differently based on this (i.e., check for a proxy and if found, establish a tunnel)

The Upgrade: header and other ws-related headers will be set by ClientWebSocket, as today.  SocketsHttpHandler should not need to detect this specifically (beyond what it does today).

@stephentoub, just to close on the last part of our convo, here's what the RFC says:

> Once a connection to the server has been established (including a
   connection via a proxy or over a TLS-encrypted tunnel), the client
   MUST send an opening handshake to the server.  The handshake consists
   of an HTTP Upgrade request, along with a list of required and
   optional header fields. 

So Upgrade happens on top of the tunnel (if any), and ClientWebSocket should provide this as part of the ws: request it constructs, regardless of tunnel.  (Basically ClientWebSocket can be ignorant of tunneling.)